sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java

```diff
@@ -723,84 +666,81 @@ public Duration getMaxReadTime() {
+            Throwable finallyBlockException = null;
+            try {
+              while ((getMaxNumRecords() == 0 || messages.size() < getMaxNumRecords())
+                     && Instant.now().isBefore(endTime)) {
+                int batchSize = DEFAULT_PULL_SIZE;
+                if (getMaxNumRecords() > 0) {
+                  batchSize = Math.min(batchSize, getMaxNumRecords() - messages.size());
+                }
 
-              PullResponse pullResponse =
-                  pubsubClient.projects().subscriptions().pull(subscription, pullRequest).execute();
-              List<String> ackIds = new ArrayList<>();
-              if (pullResponse.getReceivedMessages() != null) {
-                for (ReceivedMessage received : pullResponse.getReceivedMessages()) {
-                  messages.add(received.getMessage());
-                  ackIds.add(received.getAckId());
+                List<IncomingMessage> batchMessages =
+                    pubsubClient.pull(System.currentTimeMillis(), subscriptionPath, batchSize,
+                        false);
+                List<String> ackIds = new ArrayList<>();
+                for (IncomingMessage message : batchMessages) {
+                  messages.add(message);
+                  ackIds.add(message.ackId);
+                }
+                if (ackIds.size() != 0) {
+                  pubsubClient.acknowledge(subscriptionPath, ackIds);
                 }
               }
-
-              if (ackIds.size() != 0) {
-                AcknowledgeRequest ackRequest = new AcknowledgeRequest().setAckIds(ackIds);
-                pubsubClient.projects()
-                    .subscriptions()
-                    .acknowledge(subscription, ackRequest)
-                    .execute();
+            } catch (IOException e) {
+              throw new RuntimeException("Unexpected exception while reading from Pubsub: ", e);
+            } finally {
+              if (getSubscription() == null) {
+                try {
+                  pubsubClient.deleteSubscription(subscriptionPath);
+                } catch (Exception e) {
+                  finallyBlockException =
+                      new RuntimeException("Failed to delete subscription: ", e);
+                  LOG.error("Failed to delete subscription: ", e);
+                }
               }
             }
-          } catch (IOException e) {
-            throw new RuntimeException("Unexpected exception while reading from Pubsub: ", e);
-          } finally {
-            if (getTopic() != null) {
-              try {
-                pubsubClient.projects().subscriptions().delete(subscription).execute();
-              } catch (IOException e) {
-                finallyBlockException = new RuntimeException("Failed to delete subscription: ", e);
-                LOG.error("Failed to delete subscription: ", e);
-              }
+            if (finallyBlockException != null) {
+              Throwables.propagate(finallyBlockException);
```

Comment:
@kennknowles fyi

I think the current recommendation is simply:

```java
throw new RuntimeException("Some message", finallyBlockException);
```

Even if you intended to use `Throwables.propagate`, you should use `throw Throwables.propagate...` to work around some IDEs/cases.